### ISSUE TITLE
Evaluation on larger data set 

### ISSUE THREAD

**Issue by sanoje123** [nadavbra/protein_bert#86]
I am attempting to fine-tune your existing model with my dataset, consisting of 1 million sequences and 20,000 annotations (categorical data). I followed your provided demo code (ProteinBERT demo.ipynb). The model completed fine-tuning on the dataset, but I encountered an issue during evaluation (evaluate_by_len - function) on the training set – it consistently crashes with GPU out of memory error, regardless of how small the batch size is.

To troubleshoot, I systematically reduced the size of the training dataset to identify the maximum size for successful evaluation. Only when I decreased it to 150,000 sequences with a batch size of 64 did the evaluation succeeds. Each time, the error reported was GPU out of memory, occurring in the middle of the evaluation process. The GPU in use is an A100 40GB.

Do you maybe know the cause of this issue?

**Comment by nadavbra** [comment]
40GB should be more than enough to run ProteinBERT (we trained it on a 16GB GPU). The amount of GPU memory used should only depend on the batch size (and max_seq_len), it shouldn't matter how many records you have waiting in the CPU RAM (unless you move them to GPU memory prematurely).

**Comment by sanoje123** [comment]
I understand how your program should function. However, as you can see from the attached errors, that's not the case with us. Are you perhaps aware of why? I can provide the code if you need it.

**Output log:**

[2024_03_05-15:47:16] ========== pfam ==========
[2024_03_05-15:47:16] Output type: global categorical
[2024_03_05-15:47:20] 1086317 training set records, 126150 validation set records, 126161 test set records.
[2024_03_05-15:47:21] 17929 unique lebels.
[2024_03_05-15:47:21] Training set: Filtered out 20741 of 1086317 (1.9%) records of lengths exceeding 510.
[2024_03_05-15:48:22] Validation set: Filtered out 2300 of 126150 (1.8%) records of lengths exceeding 510.
[2024_03_05-15:48:28] *** Training-set performance: ***

------------------------------------------------------------------------------------------------------------------------------------

**Error log:**

kgcc-10.3.1 loaded successful
Bisheng-compiler-2.4.0-aarch64-linux loaded successful
openblas-0.3.18 loaded successful

Loading libs/openblas/0.3.18_bs2.4.0
  Loading requirement: compilers/bisheng/2.4.0
cuda-11.3.0 loaded successful
2024-03-05 15:49:29.914946: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 38234 MB memory:  -> device: 0, name: NVIDIA A100-PCIE-40GB, pci bus id: 0000:01:00.0, compute capability: 8.0
2024-03-05 15:49:29.920666: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 38234 MB memory:  -> device: 1, name: NVIDIA A100-PCIE-40GB, pci bus id: 0000:02:00.0, compute capability: 8.0
2024-03-05 15:49:29.922585: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:2 with 38234 MB memory:  -> device: 2, name: NVIDIA A100-PCIE-40GB, pci bus id: 0000:81:00.0, compute capability: 8.0
2024-03-05 15:49:29.924566: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:3 with 38234 MB memory:  -> device: 3, name: NVIDIA A100-PCIE-40GB, pci bus id: 0000:82:00.0, compute capability: 8.0
2024-03-05 15:49:39.919506: I tensorflow/stream_executor/cuda/cuda_blas.cc:1786] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.
2024-03-05 15:49:40.681480: I tensorflow/stream_executor/cuda/cuda_dnn.cc:368] Loaded cuDNN version 8201
2024-03-05 15:53:23.204995: W tensorflow/core/common_runtime/bfc_allocator.cc:462] Allocator (GPU_0_bfc) ran out of memory trying to allocate 16.00MiB (rounded to 16777216)requested by op model_2/global-attention-block1/transpose_4
If the cause is memory fragmentation maybe the environment variable 'TF_GPU_ALLOCATOR=cuda_malloc_async' will improve the situation. 
Current allocation summary follows.
Current allocation summary follows.
2024-03-05 15:53:23.207955: I tensorflow/core/common_runtime/bfc_allocator.cc:1010] BFCAllocator dump for GPU_0_bfc
2024-03-05 15:53:23.208003: I tensorflow/core/common_runtime/bfc_allocator.cc:1017] Bin (256): 	Total Chunks: 35, Chunks in use: 35. 8.8KiB allocated for chunks. 8.8KiB in use in bin. 264B client-requested in use in bin.
2024-03-05 15:53:23.208025: I tensorflow/core/common_runtime/bfc_allocator.cc:1017] Bin (512): 	Total Chunks: 48, Chunks in use: 48. 24.2KiB allocated for chunks. 24.2KiB in use in bin. 24.0KiB client-requested in use in bin.
2024-03-05 15:53:23.208040: I tensorflow/core/common_runtime/bfc_allocator.cc:1017] Bin (1024): 	Total Chunks: 1, Chunks in use: 1. 1.2KiB allocated for chunks. 1.2KiB in use in bin. 1.0KiB client-requested in use in bin.
2024-03-05 15:53:23.208053: I tensorflow/core/common_runtime/bfc_allocator.cc:1017] Bin (2048): 	Total Chunks: 37, Chunks in use: 37. 74.5KiB allocated for chunks. 74.5KiB in use in bin. 74.0KiB client-requested in use in bin.
2024-03-05 15:53:23.208067: I tensorflow/core/common_runtime/bfc_allocator.cc:1017] Bin (4096): 	Total Chunks: 0, Chunks in use: 0. 0B allocated for chunks. 0B in use in bin. 0B client-requested in use in bin.
2024-03-05 15:53:23.208081: I tensorflow/core/common_runtime/bfc_allocator.cc:1017] Bin (8192): 	Total Chunks: 1, Chunks in use: 1. 13.0KiB allocated for chunks. 13.0KiB in use in bin. 13.0KiB client-requested in use in bin.
2024-03-05 15:53:23.208095: I tensorflow/core/common_runtime/bfc_allocator.cc:1017] Bin (16384): 	Total Chunks: 2, Chunks in use: 1. 47.8KiB allocated for chunks. 23.8KiB in use in bin. 13.0KiB client-requested in use in bin.
.
.
.
2024-03-05 15:53:23.366599: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 2914304 totalling 2.78MiB
2024-03-05 15:53:23.366614: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 277 Chunks of size 3141632 totalling 829.92MiB
2024-03-05 15:53:23.366635: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 3163136 totalling 3.02MiB
2024-03-05 15:53:23.366650: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 3740672 totalling 3.57MiB
2024-03-05 15:53:23.366664: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 3751936 totalling 3.58MiB
2024-03-05 15:53:23.366677: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 3766272 totalling 3.59MiB
2024-03-05 15:53:23.366690: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 4059136 totalling 3.87MiB
2024-03-05 15:53:23.366705: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 333 Chunks of size 4286464 totalling 1.33GiB
2024-03-05 15:53:23.366720: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 2 Chunks of size 8388608 totalling 16.00MiB
2024-03-05 15:53:23.366734: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 8437760 totalling 8.05MiB
2024-03-05 15:53:23.366750: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 16777216 totalling 16.00MiB
2024-03-05 15:53:23.366765: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 18315264 totalling 17.47MiB
2024-03-05 15:53:23.366779: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 24602624 totalling 23.46MiB
2024-03-05 15:53:23.366794: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 33554432 totalling 32.00MiB
2024-03-05 15:53:23.366808: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 66537728 totalling 63.46MiB
2024-03-05 15:53:23.366821: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 1118697984 totalling 1.04GiB
2024-03-05 15:53:23.366836: I tensorflow/core/common_runtime/bfc_allocator.cc:1074] 1 Chunks of size 9529446400 totalling 8.87GiB
2024-03-05 15:53:23.366851: I tensorflow/core/common_runtime/bfc_allocator.cc:1078] Sum Total of in-use chunks: 37.33GiB
2024-03-05 15:53:23.366867: I tensorflow/core/common_runtime/bfc_allocator.cc:1080] total_region_allocated_bytes_: 40091779072 memory_limit_: 40091779072 available bytes: 0 curr_region_allocation_bytes_: 80183558144
2024-03-05 15:53:23.366888: I tensorflow/core/common_runtime/bfc_allocator.cc:1086] Stats: 
Limit:                     40091779072
InUse:                     40080945152
MaxInUse:                  40080945152
NumAllocs:                     5143457
MaxAllocSize:               9529446400
Reserved:                            0
PeakReserved:                        0
LargestFreeBlock:                    0

2024-03-05 15:53:23.367198: W tensorflow/core/common_runtime/bfc_allocator.cc:474] ****************************************************************************************************
2024-03-05 15:53:23.367308: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at transpose_op.cc:183 : RESOURCE_EXHAUSTED: OOM when allocating tensor with shape[32,4,64,512] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
Traceback (most recent call last):
  File "/home/share/huadjyin/home/vladimir/protein_bert/run_proteinbert.py", line 134, in <module>
    model_generator, results, confusion_matrix = run_benchmark(benchmark_name, pretrained_model_generator, input_encoder, pretraining_model_manipulation_function = \
  File "/home/share/huadjyin/home/vladimir/protein_bert/run_proteinbert.py", line 101, in run_benchmark
    results, confusion_matrix = evaluate_by_len(model_generator, input_encoder, output_spec, dataset['seq'], dataset['label'], \
  File "/home/share/huadjyin/home/vladimir/protein_bert/proteinbert/finetuning.py", line 99, in evaluate_by_len
    y_pred = model.predict(X, batch_size = batch_size)
  File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/utils/traceback_utils.py", line 67, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/tensorflow/python/eager/execute.py", line 54, in quick_execute
    tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,
tensorflow.python.framework.errors_impl.ResourceExhaustedError: Graph execution error:

Detected at node 'model_2/global-attention-block1/transpose_4' defined at (most recent call last):
    File "/home/share/huadjyin/home/vladimir/protein_bert/run_proteinbert.py", line 134, in <module>
      model_generator, results, confusion_matrix = run_benchmark(benchmark_name, pretrained_model_generator, input_encoder, pretraining_model_manipulation_function = \
    File "/home/share/huadjyin/home/vladimir/protein_bert/run_proteinbert.py", line 101, in run_benchmark
      results, confusion_matrix = evaluate_by_len(model_generator, input_encoder, output_spec, dataset['seq'], dataset['label'], \
    File "/home/share/huadjyin/home/vladimir/protein_bert/proteinbert/finetuning.py", line 99, in evaluate_by_len
      y_pred = model.predict(X, batch_size = batch_size)
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/utils/traceback_utils.py", line 64, in error_handler
      return fn(*args, **kwargs)
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/engine/training.py", line 1982, in predict
      tmp_batch_outputs = self.predict_function(iterator)
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/engine/training.py", line 1801, in predict_function
      return step_function(self, iterator)
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/engine/training.py", line 1790, in step_function
      outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/engine/training.py", line 1783, in run_step
      outputs = model.predict_step(data)
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/engine/training.py", line 1751, in predict_step
      return self(x, training=False)
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/utils/traceback_utils.py", line 64, in error_handler
      return fn(*args, **kwargs)
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/engine/base_layer.py", line 1096, in __call__
      outputs = call_fn(inputs, *args, **kwargs)
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/utils/traceback_utils.py", line 92, in error_handler
      return fn(*args, **kwargs)
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/engine/functional.py", line 451, in call
      return self._run_internal_graph(
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/engine/functional.py", line 589, in _run_internal_graph
      outputs = node.layer(*args, **kwargs)
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/utils/traceback_utils.py", line 64, in error_handler
      return fn(*args, **kwargs)
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/engine/base_layer.py", line 1096, in __call__
      outputs = call_fn(inputs, *args, **kwargs)
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/utils/traceback_utils.py", line 92, in error_handler
      return fn(*args, **kwargs)
    File "/home/share/huadjyin/home/vladimir/protein_bert/proteinbert/conv_and_global_attention_model.py", line 56, in call
      Z_batched_heads = self.calculate_attention(inputs)
    File "/home/share/huadjyin/home/vladimir/protein_bert/proteinbert/conv_and_global_attention_model.py", line 77, in calculate_attention
      KS = K.permute_dimensions(K.tanh(K.dot(S, self.Wk)), (0, 2, 3, 1))
    File "/home/share/huadjyin/home/vladimir/.conda/envs/tf_gpu/lib/python3.9/site-packages/keras/backend.py", line 3377, in permute_dimensions
      return tf.compat.v1.transpose(x, perm=pattern)
Node: 'model_2/global-attention-block1/transpose_4'
OOM when allocating tensor with shape[32,4,64,512] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
	 [[{{node model_2/global-attention-block1/transpose_4}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info. This isn't available when running in Eager mode.
 [Op:__inference_predict_function_6492]

**Comment by nadavbra** [comment]
I'm not sure what's causing this.
Maybe try to run this interactively (e.g. in Jupyter notebook) and see at what stage you run out of memory (you can use the `nvidia-smi` command to track your GPU).

**Comment by david-arredondo** [comment]
I'm having the same issue. I am only doing inference. No matter what batch size or seq_len I initialize the model with, it takes up the same 76GB on the GPU, and will crash if I try to run more than a few hundred samples with batch size 32.

**Comment by nadavbra** [comment]
At what stage of the process does your GPU memory start running out?

**Comment by david-arredondo** [comment]
The behavior has been unpredictable. I have gotten out of memory at line 71 in conv_and_global_attention
```  QX = K.tanh(K.dot(X, self.Wq)) ```

I have gotten successful embeddings for seq_len=2048, batch size = 25, 1000 samples, but it takes a long time.

Right now to test, I have the model running for seq_len=512, batch size = 5, and 10 samples. The model is taking 75 GB on an H100, and has been running for 7 minutes. Eventually I expect to get:
```
2024-07-09 14:12:42.115953: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
2024-07-09 14:12:42.132777: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 2400075000 Hz
2024-07-09 14:12:42.987781: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2024-07-09 14:15:28.891878: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.11
2024-07-09 14:15:28.900521: I tensorflow/stream_executor/cuda/cuda_blas.cc:1838] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.
2024-07-09 14:15:28.901160: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2024-07-09 14:28:41.410305: W tensorflow/core/framework/op_kernel.cc:1763] OP_REQUIRES failed at cwise_op_gpu_base.cc:89 : Internal: Failed to load in-memory CUBIN: CUDA_ERROR_NO_BINARY_FOR_GPU: no kernel image is available for execution on the device
```

Do you have a .yml file that I can use to build the correct environment? The packages suggested in the readme seem to be incorrect. Running setup.py gives issues with package compatibility, but I opened and closed an issue (please see #95) describing how I got tensorflow to recognize the GPU, but now I am having issues actually running the model.

**Comment by david-arredondo** [comment]
After some more troubleshooting, I found the following .yml to create a working conda env. The model is taking 59GB (with any seq_len and batch_size), which is more than expected but the model isn't crashing and is performing as expected.

name: proteinBERTcuda
channels:
  - conda-forge
  - defaults
dependencies:
  - pandas=1.2.3
  - cudatoolkit=11.0
  - ipykernel
  - python=3.7
  - pip:
      - tensorflow==2.4.0
      - h5py==2.10.0 #3.2.1 
      - lxml==4.3.2
      - numpy==1.19.2 #1.20.1
      - pyfaidx==0.5.8
      - tensorflow-addons==0.12.1
      - nvidia-cudnn-cu11 #
prefix: <your conda path>
